subsys/usb/class/netusb/function_rndis.c

```diff
@@ -0,0 +1,1000 @@
+	rsp->type = sys_cpu_to_le32(RNDIS_CMD_INITIALIZE_COMPLETE);
+	rsp->len = sys_cpu_to_le32(sizeof(*rsp));
+	rsp->req_id = cmd->req_id;
+
+	rsp->major_ver = sys_cpu_to_le32(RNDIS_MAJOR_VERSION);
+	rsp->minor_ver = sys_cpu_to_le32(RNDIS_MINOR_VERSION);
+
+	rsp->flags = sys_cpu_to_le32(RNDIS_FLAG_CONNECTIONLESS);
+	rsp->medium = sys_cpu_to_le32(RNDIS_MEDIUM_WIRED_ETHERNET);
+	rsp->max_packets = sys_cpu_to_le32(1);
+	rsp->max_transfer_size = sys_cpu_to_le32(rndis.mtu +
+						 sizeof(struct net_eth_hdr) +
+						 44 + 22);
+
+	rsp->pkt_align_factor = sys_cpu_to_le32(0);
+	memset(rsp->__reserved, 0, sizeof(rsp->__reserved));
+
+	rndis.state = INITIALIZED;
+
+	rndis_queue_rsp(buf);
+
+	/* Nofity about ready reply */
+	rndis_notify_rsp();
+
+	return 0;
+}
+
+static int rndis_halt_handle(void)
+{
+	SYS_LOG_DBG("");
+
+	rndis.state = UNINITIALIZED;
+
+	/* TODO: Stop networking */
+
+	return 0;
+}
+
+static uint32_t rndis_query_add_supp_list(struct net_buf *buf)
+{
+	for (int i = 0; i < ARRAY_SIZE(object_id_supported); i++) {
+		net_buf_add_le32(buf, object_id_supported[i]);
+	}
+
+	return sizeof(object_id_supported);
+}
+
+static int rndis_query_handle(u8_t *data, u32_t len)
+{
+	struct rndis_query_cmd *cmd = (void *)data;
+	struct rndis_query_cmd_complete *rsp;
+	struct net_buf *buf;
+	u32_t object_id, buf_len = 0;
+
+	buf = net_buf_alloc(&rndis_tx_pool, K_NO_WAIT);
+	if (!buf) {
+		SYS_LOG_ERR("Cannot get free buffer");
+		return -ENOMEM;
+	}
+
+	object_id = sys_le32_to_cpu(cmd->object_id);
+
+	SYS_LOG_DBG("req_id 0x%x Object ID 0x%x buf_len %u buf_offset %u",
+		    sys_le32_to_cpu(cmd->req_id),
+		    object_id,
+		    sys_le32_to_cpu(cmd->buf_len),
+		    sys_le32_to_cpu(cmd->buf_offset));
+
+	rsp = net_buf_add(buf, sizeof(*rsp));
+	rsp->type = sys_cpu_to_le32(RNDIS_CMD_QUERY_COMPLETE);
+	rsp->req_id = cmd->req_id;
+
+	/* offset is from the beginning of the req_id field */
+	rsp->buf_offset = sys_cpu_to_le32(16);
+
+	switch (object_id) {
+	case RNDIS_OBJECT_ID_GEN_SUPP_LIST:
+		SYS_LOG_DBG("RNDIS_OBJECT_ID_GEN_SUPP_LIST");
+		rndis_query_add_supp_list(buf);
+		break;
+	case RNDIS_OBJECT_ID_GEN_PHYSICAL_MEDIUM:
+		SYS_LOG_DBG("RNDIS_OBJECT_ID_GEN_PHYSICAL_MEDIUM");
+		net_buf_add_le32(buf, RNDIS_PHYSICAL_MEDIUM_TYPE_UNSPECIFIED);
+		break;
+	case RNDIS_OBJECT_ID_GEN_MAX_FRAME_SIZE:
+		SYS_LOG_DBG("RNDIS_OBJECT_ID_GEN_MAX_FRAME_SIZE");
+		net_buf_add_le32(buf, rndis.mtu);
+		break;
+	case RNDIS_OBJECT_ID_GEN_LINK_SPEED:
+		SYS_LOG_DBG("RNDIS_OBJECT_ID_GEN_LINK_SPEED");
+		if (rndis.media_status == RNDIS_OBJECT_ID_MEDIA_DISCONNECTED) {
+			net_buf_add_le32(buf, 0);
+		} else {
+			net_buf_add_le32(buf, rndis.speed);
+		}
+		break;
+	case RNDIS_OBJECT_ID_GEN_CONN_MEDIA_STATUS:
+		SYS_LOG_DBG("RNDIS_OBJECT_ID_GEN_CONN_MEDIA_STATUS");
+		net_buf_add_le32(buf, rndis.media_status);
+		break;
+	case RNDIS_OBJECT_ID_GEN_MAX_TOTAL_SIZE:
+		SYS_LOG_DBG("RNDIS_OBJECT_ID_GEN_MAX_TOTAL_SIZE");
+		net_buf_add_le32(buf, 1558);
```

Comment:
What is this magic value?

Reply:
This is maximum total packet length 
https://docs.microsoft.com/en-us/windows-hardware/drivers/network/oid-gen-maximum-total-size